sound/soc/intel/boards/skl_hda_dsp_generic.c

```diff
@@ -135,11 +138,20 @@ static int skl_hda_fill_card_info(struct snd_soc_acpi_mach_params *mach_params)
 				skl_hda_be_dai_links[IDISP_DAI_COUNT +
 					HDAC_DAI_COUNT + i];
 		}
-	} else if (codec_count == 2 && codec_mask & IDISP_CODEC_MASK) {
+	} else if (codec_count <= 2) {
 		num_links = ARRAY_SIZE(skl_hda_be_dai_links);
 		num_route = ARRAY_SIZE(skl_hda_map);
 		card->dapm_widgets = skl_hda_widgets;
 		card->num_dapm_widgets = ARRAY_SIZE(skl_hda_widgets);
+		if (!(codec_mask & IDISP_CODEC_MASK)) {
+			if (codec_count == 2)
+				return -EINVAL;
+			for (i = 0; i < IDISP_DAI_COUNT; i++) {
+				skl_hda_be_dai_links[i].codecs = dummy_codec;
+				skl_hda_be_dai_links[i].num_codecs =
+					ARRAY_SIZE(dummy_codec);
+			}
+		}
```

Comment:
you should clarify what happens for other machine drivers? Why would we ignore HDMI only in the case of the HDaudio codec cases?

Reply:
@plbossart We probably should modify other machine drivers, but that is less critical as those are for specific devices where HDMI either is or is not part of the product configuration, so there is less expected variation. We could either go ahead and change all the other machien drivers as well, or mybe wait for @ranj063 's mfd work to reach the point where hdmi is a separate client and potentially easier to isolate.

But for the generic HDA driver, this is much more urgent. The hda generic driver is our catch-all driver that should be much more robust against variations in configuration than the other more specific machine drivers. E.g. any machine that has IGFX disabled and has DMICs, will just fail out-of-the-box with upstream kernel now without this patch.

Reply:
@kv2019i this logic seems twisted. If idisp codecs are disabled, loading a topology with idisp pipelines should fail IMO.
By loading the tplg with idsp pipelines and using dummy codecs, we're giving the wrong signal

Reply:
Alternativey, what we should do is fail gracefully ie not let the card registration fail entirely but somehow ignore the idisp pipelines in tplg with a warning

Reply:
@ranj063 we don't have a good way to discard parts of the topology when the hardware described in those topology files is just not present.

I think @kv2019i choses the least invasive solution, the pipelines remain but are essentially never used since the jack detection will never fail. However there could still be a problem if someone decided to use aplay -Dhw:x,y. @kv2019i what would happen then? Underflow?

Reply:
@plbossart @ranj063 Ack, exactly. If i915 fails to initialize and you try to aplay to one of the PCMs, you will just get an error from failing hw_error and playback won't start. Due to inherent hotplug nature of HDMI/DP monitors, no application should assume that existance of PCMs indicates that a monitor is connected. The routing is not guaranteed, so this is just plain wrong logic.

If we go with the approach of using different topologies, then we end up maintaining three topologies for each platform (hdmi and some codec, only hdmi and only some codec). If we have similar availability concerns with some other codecs, do we then provide topologies for all combinations (codec1, codec2; codec1 only; codec 2 only; and all combinations of codec1 and codec2 with and without hdmi). 

I think a simpler approach is that if only difference in topology is that some of the edge leafs of the graph are not available, we are better of using a single topology and just handling enable/disable for these leaf nodes (like HDMI here). The leafs could be input or outputs. Only if the difference is such, that we would set up a different looking topology altogether (and save some DSP resource), then maintaining it separately makes sense.